src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,169 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int BASE_ARRAY_LENGTH = 10;
+    private static final double ARRAY_INCREMENT_VALUE = 1.5;
+    private T[] array;
+    private int currentSlot;
+
+    public ArrayList() {
+        array = (T[]) new Object[BASE_ARRAY_LENGTH];
+        currentSlot = -1;
+    }
+
     @Override
     public void add(T value) {
-
+        checkAndIncreaseArrayLength();
+        currentSlot++;
+        array[currentSlot] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > currentSlot + 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        checkAndIncreaseArrayLength();
+        currentSlot++;
+        for (int i = size() - 1; i >= (index == 0 ? index + 1 : index); i--) {
+            array[i] = array[i - 1];
+        }
+        array[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkAndIncreaseArrayLength(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            currentSlot++;
+            array[currentSlot] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds");
+        }
+        T date = array[index];
+        for (int i = index; i < size(); i++) {
+            if (index == currentSlot) {
+                array[index] = null;
+                currentSlot--;
+                return date;
+            }
+            array[i] = array[i + 1];
+        }
+        currentSlot--;
+        array[size()] = null;
+        return date;
```

Comment:
what if we remove an element in the middle and there are some elements after it? Will it work correctly?
Use `System.arraycopy()` to move the elements after it